File: system_monitor.go
// system_monitor.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"runtime"
	"strconv"
	"strings"
	"time"
)

type Monitor struct {
	interval       int
	running        bool
	logging        bool
	showProcesses  bool
}

func NewMonitor(interval int) *Monitor {
	return &Monitor{
		interval: interval,
		running:  true,
	}
}

func getCPUUsage() (float64, error) {
	if runtime.GOOS == "windows" {
		// Using wmic to get CPU load
		cmd := exec.Command("wmic", "cpu", "get", "loadpercentage")
		out, err := cmd.Output()
		if err != nil {
			return 0, err
		}
		lines := strings.Split(string(out), "\n")
		for _, line := range lines {
			line = strings.TrimSpace(line)
			if line != "" && line != "LoadPercentage" {
				val, err := strconv.ParseFloat(line, 64)
				if err == nil {
					return val, nil
				}
			}
		}
		return 0, nil
	} else {
		// Unix: use ps aux or /proc/stat (simplified: use ps for demo)
		// More accurate would be to read /proc/stat, but for simplicity we use 'ps' or 'top'
		cmd := exec.Command("sh", "-c", "top -bn1 | grep 'Cpu(s)' | awk '{print $2}' | cut -d'%' -f1")
		out, err := cmd.Output()
		if err != nil {
			return 0, err
		}
		val, err := strconv.ParseFloat(strings.TrimSpace(string(out)), 64)
		if err != nil {
			return 0, err
		}
		return val, nil
	}
}

func getMemoryUsage() (total, used, free uint64, percent float64) {
	if runtime.GOOS == "windows" {
		// Windows: use wmic os get TotalVisibleMemorySize,FreePhysicalMemory
		cmd := exec.Command("wmic", "os", "get", "TotalVisibleMemorySize,FreePhysicalMemory")
		out, err := cmd.Output()
		if err != nil {
			return 0, 0, 0, 0
		}
		lines := strings.Split(string(out), "\n")
		for _, line := range lines {
			line = strings.TrimSpace(line)
			if line == "" || strings.HasPrefix(line, "TotalVisibleMemorySize") {
				continue
			}
			fields := strings.Fields(line)
			if len(fields) >= 2 {
				totalKB, _ := strconv.ParseUint(fields[0], 10, 64)
				freeKB, _ := strconv.ParseUint(fields[1], 10, 64)
				total = totalKB * 1024
				free = freeKB * 1024
				used = total - free
				percent = float64(used) / float64(total) * 100
				return
			}
		}
	} else {
		// Unix: read /proc/meminfo
		data, err := os.ReadFile("/proc/meminfo")
		if err != nil {
			return 0, 0, 0, 0
		}
		lines := strings.Split(string(data), "\n")
		var memTotal, memAvailable uint64
		for _, line := range lines {
			line = strings.TrimSpace(line)
			if strings.HasPrefix(line, "MemTotal:") {
				fields := strings.Fields(line)
				if len(fields) >= 2 {
					memTotal, _ = strconv.ParseUint(fields[1], 10, 64)
					memTotal *= 1024 // KB to bytes
				}
			} else if strings.HasPrefix(line, "MemAvailable:") {
				fields := strings.Fields(line)
				if len(fields) >= 2 {
					memAvailable, _ = strconv.ParseUint(fields[1], 10, 64)
					memAvailable *= 1024
				}
			}
		}
		if memTotal > 0 {
			total = memTotal
			free = memAvailable
			used = total - free
			percent = float64(used) / float64(total) * 100
		}
	}
	return
}

func formatBar(percent float64, width int) string {
	filled := int(percent / 100 * float64(width))
	bar := strings.Repeat("█", filled) + strings.Repeat("░", width-filled)
	return bar
}

func (m *Monitor) render() {
	cpu, _ := getCPUUsage()
	total, used, free, percent := getMemoryUsage()
	totalGB := float64(total) / (1024 * 1024 * 1024)
	usedGB := float64(used) / (1024 * 1024 * 1024)

	fmt.Print("\033[H\033[2J") // clear screen
	fmt.Println("=== System Monitor ===")
	fmt.Printf("Refresh interval: %ds (press 'q' to quit, '+'/'-' to adjust, 'l' log, 'p' processes)\n\n", m.interval)
	cpuBar := formatBar(cpu, 30)
	memBar := formatBar(percent, 30)
	fmt.Printf("CPU:   %s  %5.1f%%\n", cpuBar, cpu)
	fmt.Printf("RAM:   %s  %5.1f%%  (Used: %.1f GB / Total: %.1f GB)\n", memBar, percent, usedGB, totalGB)

	if m.showProcesses {
		fmt.Println("\nTop processes by CPU:")
		// Placeholder: we won't implement full process list in Go for brevity
		fmt.Println("(Process list not implemented in this version)")
	}
}

func (m *Monitor) run() {
	fmt.Println("Starting System Monitor... (press 'q' to quit)")
	go func() {
		// Listen for keypresses (simple: just read from stdin)
		var b []byte = make([]byte, 1)
		for m.running {
			os.Stdin.Read(b)
			switch b[0] {
			case 'q':
				m.running = false
			case '+':
				if m.interval < 10 {
					m.interval++
				}
			case '-':
				if m.interval > 1 {
					m.interval--
				}
			case 'l':
				m.logging = !m.logging
				fmt.Println("\nLogging toggled.")
			case 'p':
				m.showProcesses = !m.showProcesses
			}
		}
	}()

	for m.running {
		m.render()
		time.Sleep(time.Duration(m.interval) * time.Second)
	}
	fmt.Println("Goodbye!")
}

func main() {
	monitor := NewMonitor(2)
	monitor.run()
}

File: system_monitor.swift
// system_monitor.swift
import Foundation

class SystemMonitor {
    var interval: Int
    var running: Bool
    var logging: Bool
    var showProcesses: Bool
    var logFile: FileHandle?

    init(interval: Int = 2) {
        self.interval = interval
        self.running = true
        self.logging = false
        self.showProcesses = false
    }

    func getCPUUsage() -> Double {
        // Use system commands (simplified)
        let task = Process()
        task.executableURL = URL(fileURLWithPath: "/usr/bin/env")
        task.arguments = ["sh", "-c", "top -l1 | grep 'CPU usage' | awk '{print $3}' | cut -d'%' -f1"]
        let pipe = Pipe()
        task.standardOutput = pipe
        task.standardError = FileHandle.nullDevice
        try? task.run()
        task.waitUntilExit()
        let data = pipe.fileHandleForReading.readDataToEndOfFile()
        guard let output = String(data: data, encoding: .utf8) else { return 0 }
        if let val = Double(output.trimmingCharacters(in: .whitespacesAndNewlines)) {
            return val
        }
        return 0
    }

    func getMemoryUsage() -> (totalGB: Double, usedGB: Double, percent: Double) {
        // Use sysctl for total, vm_stat for free
        var total: UInt64 = 0
        var size = MemoryLayout<UInt64>.size
        if sysctlbyname("hw.memsize", &total, &size, nil, 0) == 0 {
            let totalGB = Double(total) / 1_073_741_824.0
            // Try to get free memory via vm_stat
            let task = Process()
            task.executableURL = URL(fileURLWithPath: "/usr/bin/vm_stat")
            let pipe = Pipe()
            task.standardOutput = pipe
            task.standardError = FileHandle.nullDevice
            try? task.run()
            task.waitUntilExit()
            let data = pipe.fileHandleForReading.readDataToEndOfFile()
            guard let output = String(data: data, encoding: .utf8) else { return (0,0,0) }
            let lines = output.split(separator: "\n")
            var freePages: UInt64 = 0
            for line in lines {
                if line.contains("Pages free:") {
                    let parts = line.split(separator: ":")
                    if parts.count > 1 {
                        freePages = UInt64(parts[1].trimmingCharacters(in: .whitespaces)) ?? 0
                        break
                    }
                }
            }
            let pageSize = sysconf( _SC_PAGESIZE )
            let freeBytes = UInt64(freePages) * UInt64(pageSize)
            let usedBytes = total - freeBytes
            let usedGB = Double(usedBytes) / 1_073_741_824.0
            let percent = (Double(usedBytes) / Double(total)) * 100
            return (totalGB, usedGB, percent)
        }
        return (0,0,0)
    }

    func formatBar(percent: Double, width: Int = 30) -> String {
        let filled = Int((percent / 100.0) * Double(width))
        return String(repeating: "█", count: filled) + String(repeating: "░", count: width - filled)
    }

    func render() {
        let cpu = getCPUUsage()
        let mem = getMemoryUsage()
        let cpuBar = formatBar(percent: cpu)
        let memBar = formatBar(percent: mem.percent)
        // Clear screen
        print("\u{001B}[2J\u{001B}[H", terminator: "")
        print("=== System Monitor ===")
        print("Refresh interval: \(interval)s (press 'q' to quit, '+'/'-' to adjust, 'l' log, 'p' processes)")
        print()
        print("CPU:   \(cpuBar)  \(String(format: "%.1f", cpu))%")
        print("RAM:   \(memBar)  \(String(format: "%.1f", mem.percent))%  (Used: \(String(format: "%.1f", mem.usedGB)) GB / Total: \(String(format: "%.1f", mem.totalGB)) GB)")

        if showProcesses {
            print("\nTop processes by CPU:")
            // Placeholder
            print("(Process list not implemented in this version)")
        }

        if logging {
            if logFile == nil {
                let path = FileManager.default.currentDirectoryPath + "/system_monitor.csv"
                if !FileManager.default.fileExists(atPath: path) {
                    try? "timestamp,cpu,mem_percent,mem_used_gb,mem_total_gb\n".write(toFile: path, atomically: true, encoding: .utf8)
                }
                logFile = FileHandle(forWritingAtPath: path)
                logFile?.seekToEndOfFile()
            }
            let line = "\(Date().iso8601String()),\(String(format: "%.1f", cpu)),\(String(format: "%.1f", mem.percent)),\(String(format: "%.2f", mem.usedGB)),\(String(format: "%.2f", mem.totalGB))\n"
            if let data = line.data(using: .utf8) {
                logFile?.write(data)
            }
        }
    }

    func run() {
        print("Starting System Monitor... (press q to quit)")
        // Set up stdin to not be line-buffered
        var oldTerm = termios()
        tcgetattr(STDIN_FILENO, &oldTerm)
        var newTerm = oldTerm
        newTerm.c_lflag &= ~UInt(ICANON)
        newTerm.c_lflag &= ~UInt(ECHO)
        tcsetattr(STDIN_FILENO, TCSANOW, &newTerm)

        DispatchQueue.global().async {
            while self.running {
                var ch: UInt8 = 0
                read(STDIN_FILENO, &ch, 1)
                let char = Character(UnicodeScalar(ch))
                switch char {
                case "q": self.running = false
                case "+": self.interval = min(self.interval + 1, 10)
                case "-": self.interval = max(self.interval - 1, 1)
                case "l": self.logging.toggle(); print("\nLogging toggled.")
                case "p": self.showProcesses.toggle()
                default: break
                }
            }
        }

        while running {
            render()
            Thread.sleep(forTimeInterval: TimeInterval(interval))
        }
        logFile?.closeFile()
        print("\nGoodbye!")
    }
}

extension Date {
    func iso8601String() -> String {
        let formatter = ISO8601DateFormatter()
        return formatter.string(from: self)
    }
}

SystemMonitor().run()
